lightning/src/ln/channel.rs

```diff
@@ -2518,6 +2523,52 @@ struct CommitmentTxInfoCached {
 	feerate: u32,
 }
 
+/// Contents of a wire message that fails an HTLC backwards. Useful for [`Channel::fail_htlc`] to
+/// fail with either [`msgs::UpdateFailMalformedHTLC`] or [`msgs::UpdateFailHTLC`] as needed.
+pub(super) trait FailHTLCContents {
+	type Message;
+	fn message(self, htlc_id: u64, channel_id: ChannelId) -> Self::Message;
+	fn message_name() -> &'static str;
+	fn inbound_htlc_state(self) -> InboundHTLCState;
+	fn htlc_update_awaiting_ack(self, htlc_id: u64) -> HTLCUpdateAwaitingACK;
+}
+impl FailHTLCContents for msgs::OnionErrorPacket {
+	type Message = msgs::UpdateFailHTLC;
+	fn message(self, htlc_id: u64, channel_id: ChannelId) -> Self::Message {
+		msgs::UpdateFailHTLC { htlc_id, channel_id, reason: self }
+	}
+	fn message_name() -> &'static str { "update_fail_htlc" }
+	fn inbound_htlc_state(self) -> InboundHTLCState {
+		InboundHTLCState::LocalRemoved(InboundHTLCRemovalReason::FailRelay(self))
+	}
+	fn htlc_update_awaiting_ack(self, htlc_id: u64) -> HTLCUpdateAwaitingACK {
+		HTLCUpdateAwaitingACK::FailHTLC { htlc_id, err_packet: self }
+	}
+}
+impl FailHTLCContents for (u16, [u8; 32]) {
+	type Message = msgs::UpdateFailMalformedHTLC; // (failure_code, sha256_of_onion)
+	fn message(self, htlc_id: u64, channel_id: ChannelId) -> Self::Message {
+		let (failure_code, sha256_of_onion) = self;
+		msgs::UpdateFailMalformedHTLC {
+			htlc_id,
+			channel_id,
+			sha256_of_onion,
+			failure_code,
+		}
+	}
+	fn message_name() -> &'static str { "update_fail_malformed_htlc" }
+	fn inbound_htlc_state(self) -> InboundHTLCState {
+		let (failure_code, sha256_of_onion) = self;
```

Comment:
Should we normalize the order so this doesn't need to be destructured?

Reply:
Got rid of the destructuring. Not sure I got what you meant by "normalize."

Reply:
By that I meant to keep the order of `(failure_code, sha256_of_onion)` the same between here and in `InboundHTLCRemovalReason::FailMalformed`. That way you just need to wrap `self` here rather than accessing each item in the tuple.

Reply:
Oh, okay. I'll pass on this for now but I'll do it if there ends up being a follow-up to this PR.